### PR TITLE
Auto-Join: Configurable Scheme & Port (and add k8s provider)

### PR DESCRIPTION
**Pull request by alexanderbez**
- Allow operators to define a scheme and/or port in the `retry_join` stanza used for auto-join IP discovery. Vault defaults to HTTPS/8200 when these values are not provided.
- Introduce k8s provider (tested by @jasonodonnell)
- Update docs

### CODE REVIEW

File: vault/raft.go [hunk 1]
```diff
@@ -940,6 +940,32 @@ func (c *Core) JoinRaftCluster(ctx context.Context, leaderInfos []*raft.LeaderJo
...
+					}
+
+					if u.Scheme == "" {
+						scheme := leaderInfo.Scheme
+						if scheme == "" {
+							// default to HTTPS when no scheme is provided
+							scheme = "https"
+						}
+
+						addr = fmt.Sprintf("%s://%s", scheme, addr)
```
Comment (ncabatoff) [inline vault/raft.go:956]: Why not just populate `u`'s members, then call `String()` on it?

Comment (alexanderbez) [reply vault/raft.go:956]: We can also do that :)

**Review by ncabatoff — APPROVED** [review]

File: physical/raft/raft.go [hunk 2]
```diff
@@ -130,6 +130,13 @@ type LeaderJoinInfo struct {
 	// via 'leader_api_addr'.
 	AutoJoin string `json:"auto_join"`
 
+	// Scheme defines the optional URI protocol scheme for addresses discovered via
+	// auto-join.
+	Scheme string `json:"scheme"`
```
Comment (ncabatoff) [inline physical/raft/raft.go:135]: Do we want to include AutoJoin in the name to make explicit what it's for?

Comment (alexanderbez) [reply physical/raft/raft.go:135]: Yeah I was thinking we should.

File: physical/raft/raft.go [hunk 3]
```diff
@@ -130,6 +130,14 @@ type LeaderJoinInfo struct {
 	// via 'leader_api_addr'.
 	AutoJoin string `json:"auto_join"`
 
+	// AutoJoinScheme defines the optional URI protocol scheme for addresses
+	// discovered via auto-join.
+	AutoJoinScheme string `json:"auto_join_scheme"`
```
Comment (ncabatoff) [inline physical/raft/raft.go:135]: Do we want omitempty on this and the other AutoJoin fields?

Comment (alexanderbez) [reply physical/raft/raft.go:135]: Depends. Is this configuration exposed in any requests or other client-facing calls? I'm suspicion is that it's not. @briankassouf?

Comment (briankassouf) [reply physical/raft/raft.go:135]: Yeah, I don't actually think that we ever json marshal this struct, but you may want to confirm that.

File: vault/raft.go [hunk 4]
```diff
@@ -1158,6 +1185,8 @@ func newDiscover() (*discover.Discover, error) {
 		providers[k] = v
 	}
 
+	providers["k8s"] = &discoverk8s.Provider{}
```
Comment (alexanderbez) [inline vault/raft.go:1188]: /cc @jasonodonnell 👍

File: physical/raft/raft.go (hunk 3 above)
Comment (ncabatoff) [reply physical/raft/raft.go:135]: I was thinking about the sanitized config output, though perhaps it doesn't marshall this directly.

**Review by ncabatoff — APPROVED** [review]

**Comment by alexanderbez** [comment]
1.6.x backport => https://github.com/hashicorp/vault/pull/10229

**Pull request merged**